protocol/v1/qbft/instance/pre_prepare.go

```diff
@@ -48,24 +48,39 @@ func (i *Instance) prePrepareMsgValidationPipeline() pipelines.SignedMessagePipe
 // 			∀ <ROUND-CHANGE, λi, round, prj , pvj> ∈ Qrc : prj = ⊥ ∧ prj = ⊥
 // 			∨ received a quorum of valid <PREPARE, λi, pr, value> messages such that:
 // 				(pr, value) = HighestPrepared(Qrc)
-func (i *Instance) JustifyPrePrepare(round uint64, preparedValue []byte) error {
+func (i *Instance) JustifyPrePrepare(round uint64, proposalData *specqbft.ProposalData) error {
 	if round == 1 {
 		return nil
 	}
 
-	if quorum, _, _ := i.changeRoundQuorum(specqbft.Round(round)); quorum {
-		notPrepared, highest, err := i.HighestPrepared(specqbft.Round(round))
-		if err != nil {
-			return err
-		}
-		if notPrepared {
-			return nil
-		} else if !bytes.Equal(preparedValue, highest.PreparedValue) {
-			return errors.New("preparedValue different than highest prepared")
+	for _, rc := range proposalData.RoundChangeJustification {
+		if err := i.validateRoundChange(rc); err != nil {
+			return errors.Wrap(err, "change round msg not valid")
 		}
+	}
+
+	if quorum, _, _ := i.changeRoundQuorum(proposalData.RoundChangeJustification); !quorum {
+		return errors.New("no change round quorum")
+	}
+	notPrepared, highest, err := i.HighestPrepared(specqbft.Round(round))
+	if err != nil {
+		return err
+	}
+	if notPrepared {
 		return nil
 	}
-	return errors.New("no change round quorum")
+	if !bytes.Equal(proposalData.Data, highest.PreparedValue) {
+		return errors.New("preparedValue different than highest prepared")
+	}
+	return nil
+}
+
+func (i *Instance) validateRoundChange(signedMsg *specqbft.SignedMessage) error {
+	if len(signedMsg.GetSigners()) != 1 {
```

Comment:
Thats the only validation we need? (from spec?)

Reply:
@nivBlox No, but other validations were done in other places because I needed to keep the order of checks similar to spec. However, I will likely refactor some checks a little bit after finishing with other message types because I will know what checks are used in what places, so the scope of changes will be clearer.